File: client/lib/services/settings_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'server_config.dart';

class SettingsService {
  final ServerConfig _config;
  Map<String, dynamic> _settings = {};
  bool _loaded = false;

  SettingsService(this._config);

  Map<String, dynamic> get all => Map.from(_settings);
  bool get isLoaded => _loaded;

  Map<String, dynamic> _voice() =>
      Map<String, dynamic>.from(_settings['voice'] as Map? ?? {});
  Map<String, dynamic> _ui() =>
      Map<String, dynamic>.from(_settings['ui'] as Map? ?? {});
  Map<String, dynamic> _health() =>
      Map<String, dynamic>.from(_settings['health'] as Map? ?? {});
  Map<String, dynamic> _smartHome() =>
      Map<String, dynamic>.from(_settings['smart_home'] as Map? ?? {});

  bool get wakeWordEnabled => _voice()['wake_word_enabled'] as bool? ?? true;
  double get wakeWordSensitivity =>
      (_voice()['wake_word_sensitivity'] as num?)?.toDouble() ?? 0.5;
  String get ttsVoice => _voice()['tts_voice'] as String? ?? 'en-US-GuyNeural';
  String get voiceProfile => _voice()['voice_profile'] as String? ?? 'jarvis';
  String get llmModel => _voice()['llm_model'] as String? ?? 'llama3.2';
  bool get pushToTalk => _voice()['push_to_talk'] as bool? ?? false;

  bool get darkMode => _ui()['dark_mode'] as bool? ?? true;
  bool get notificationsEnabled =>
      _ui()['notifications_enabled'] as bool? ?? true;

  bool get healthAlertsEnabled =>
      _health()['alerts_enabled'] as bool? ?? false;
  bool get heartRateAlerts =>
      _health()['heart_rate_alerts'] as bool? ?? true;
  bool get spo2Alerts => _health()['spo2_alerts'] as bool? ?? true;

  String get mqttBroker =>
      _smartHome()['mqtt_broker'] as String? ?? '';
  int get mqttPort => (_smartHome()['mqtt_port'] as num?)?.toInt() ?? 1883;
  String get mqttUsername =>
      _smartHome()['mqtt_username'] as String? ?? '';
  String get mqttPassword =>
      _smartHome()['mqtt_password'] as String? ?? '';

  Map<String, String> get _authHeaders => {
    if (_config.token != null) 'Authorization': 'Bearer ${_config.token}',
    'Content-Type': 'application/json',
  };

  Future<void> fetch() async {
    final response = await http
        .get(Uri.parse('${_config.baseUrl}/api/v1/settings'), headers: _authHeaders)
        .timeout(Duration(seconds: 10));
    if (response.statusCode == 200) {
      _settings = json.decode(response.body) as Map<String, dynamic>;
      _loaded = true;
    } else {
      throw Exception('Failed to fetch settings: ${response.statusCode}');
    }
  }

  Future<void> save() async {
    final response = await http
        .put(
          Uri.parse('${_config.baseUrl}/api/v1/settings'),
          headers: _authHeaders,
          body: json.encode(_settings),
        )
        .timeout(Duration(seconds: 10));
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      _settings = data['settings'] as Map<String, dynamic>? ?? _settings;
    } else {
      throw Exception('Failed to save settings: ${response.statusCode}');
    }
  }

  Future<void> patch(Map<String, dynamic> partial) async {
    final response = await http
        .patch(
          Uri.parse('${_config.baseUrl}/api/v1/settings'),
          headers: _authHeaders,
          body: json.encode(partial),
        )
        .timeout(Duration(seconds: 10));
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      _settings = data['settings'] as Map<String, dynamic>? ?? _settings;
    } else {
      throw Exception('Failed to patch settings: ${response.statusCode}');
    }
  }
}

File: backend/app/services/settings_service.py
import json
from pathlib import Path
from threading import Lock

DATA_DIR = Path(__file__).parent.parent.parent / "data"
SETTINGS_FILE = DATA_DIR / "settings.json"

DEFAULT_SETTINGS = {
    "voice": {
        "wake_word_enabled": True,
        "wake_word_sensitivity": 0.5,
        "tts_voice": "en-US-GuyNeural",
        "voice_profile": "jarvis",
        "llm_model": "llama3.2",
        "stt_model": "tiny",
        "push_to_talk": False,
    },
    "ui": {
        "dark_mode": True,
        "notifications_enabled": True,
    },
    "health": {
        "alerts_enabled": False,
        "heart_rate_alerts": True,
        "spo2_alerts": True,
    },
    "smart_home": {
        "mqtt_broker": "",
        "mqtt_port": 1883,
        "mqtt_username": "",
        "mqtt_password": "",
    },
}


class SettingsService:
    def __init__(self):
        self._profiles: dict[str, dict] = {}
        self._lock = Lock()
        self._load()

    def _ensure_file(self):
        DATA_DIR.mkdir(parents=True, exist_ok=True)

    def _load(self):
        self._ensure_file()
        if SETTINGS_FILE.exists():
            try:
                with open(SETTINGS_FILE) as f:
                    self._profiles = json.load(f)
            except (json.JSONDecodeError, OSError):
                self._profiles = {}
        if "default" not in self._profiles:
            self._profiles["default"] = {}
            self._save()

    def _save(self):
        self._ensure_file()
        with open(SETTINGS_FILE, "w") as f:
            json.dump(self._profiles, f, indent=2)

    def _deep_merge(self, base: dict, override: dict) -> dict:
        result = dict(base)
        for key, value in override.items():
            if key in result and isinstance(result[key], dict) and isinstance(value, dict):
                result[key] = self._deep_merge(result[key], value)
            else:
                result[key] = value
        return result

    def get_all(self, profile_id: str = "default") -> dict:
        with self._lock:
            overrides = self._profiles.get(profile_id, {})
            return self._deep_merge(DEFAULT_SETTINGS, overrides)

    def get(self, profile_id: str, category: str, key: str = None):
        full = self.get_all(profile_id)
        section = full.get(category)
        if section is None:
            return None
        if key is None:
            return dict(section)
        return section.get(key)

    def update(self, profile_id: str, partial: dict) -> dict:
        with self._lock:
            current = self._profiles.get(profile_id, {})
            merged = self._deep_merge(current, partial)
            self._profiles[profile_id] = merged
            self._save()
            return self._deep_merge(DEFAULT_SETTINGS, merged)

    def reset(self, profile_id: str) -> dict:
        with self._lock:
            self._profiles.pop(profile_id, None)
            self._save()
            return dict(DEFAULT_SETTINGS)


settings_service = SettingsService()
